programs/mango-v4/src/instructions/token_force_close_borrows_with_token.rs

```diff
@@ -0,0 +1,207 @@
+
+        msg!(
+            "Force closed {} liab for {} asset",
+            liab_transfer,
+            asset_transfer
+        );
+
+        // liqee asset
+        emit!(TokenBalanceLog {
+            mango_group: liqee.fixed.group,
+            mango_account: liqee_key,
+            token_index: asset_token_index,
+            indexed_position: liqee_asset_indexed_position.to_bits(),
+            deposit_index: asset_bank.deposit_index.to_bits(),
+            borrow_index: asset_bank.borrow_index.to_bits(),
+        });
+        // liqee liab
+        emit!(TokenBalanceLog {
+            mango_group: liqee.fixed.group,
+            mango_account: liqee_key,
+            token_index: liab_token_index,
+            indexed_position: liqee_liab_indexed_position.to_bits(),
+            deposit_index: liab_bank.deposit_index.to_bits(),
+            borrow_index: liab_bank.borrow_index.to_bits(),
+        });
+        // liqor asset
+        emit!(TokenBalanceLog {
+            mango_group: liqee.fixed.group,
+            mango_account: liqor_key,
+            token_index: asset_token_index,
+            indexed_position: liqor_asset_indexed_position.to_bits(),
+            deposit_index: asset_bank.deposit_index.to_bits(),
+            borrow_index: asset_bank.borrow_index.to_bits(),
+        });
+        // liqor liab
+        emit!(TokenBalanceLog {
+            mango_group: liqee.fixed.group,
+            mango_account: liqor_key,
+            token_index: liab_token_index,
+            indexed_position: liqor_liab_indexed_position.to_bits(),
+            deposit_index: liab_bank.deposit_index.to_bits(),
+            borrow_index: liab_bank.borrow_index.to_bits(),
+        });
+
+        // TODO: hmm should we just check if init_health was improved OR
+        // should we use this opportunity to get the liqee out of liquidated state?
+        let liqee_liq_end_health = liqee_health_cache.health(HealthType::LiquidationEnd);
+        liqee
+            .fixed
+            .maybe_recover_from_being_liquidated(liqee_liq_end_health);
+
+        // Since we use a scanning account retriever, it's safe to deactivate inactive token positions
+        if !liqee_asset_active {
+            liqee.deactivate_token_position_and_log(liqee_asset_raw_index, liqee_key);
+        }
+        if !liqee_liab_active {
+            liqee.deactivate_token_position_and_log(liqee_liab_raw_index, liqee_key);
+        }
+        if !liqor_asset_active {
+            liqor.deactivate_token_position_and_log(liqor_asset_raw_index, liqor_key);
+        }
+        if !liqor_liab_active {
+            liqor.deactivate_token_position_and_log(liqor_liab_raw_index, liqor_key)
+        }
+    };
+
+    // Check liqor's health
+    let liqor_health = compute_health(&liqor.borrow(), HealthType::Init, &account_retriever)
+        .context("compute liqor health")?;
+    require!(liqor_health >= 0, MangoError::HealthMustBePositive);
```

Comment:
technically this should always improve liqor health, since we decrease the zero-asset-weight liab token and gain some asset token. but this check can stay for safety -- maybe leave a comment?